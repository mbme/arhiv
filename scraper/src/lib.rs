#![deny(clippy::all)]

use std::collections::HashMap;

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};

use rs_utils::{log, run_command_with_envs, Chromium, NodeJS, TempFile};

fn get_script_temp_file() -> Result<TempFile> {
    let script = include_str!("../dist/node-scraper.js");

    // TODO use "shared memory file" shm_open
    let temp_file = TempFile::new_with_details("scrape-script-", ".js");

    temp_file.write(script)?;

    Ok(temp_file)
}

#[derive(Serialize, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ScrapeResult {
    url: String,
    original_url: Option<String>,
    scraper_name: Option<String>,
    data: Option<serde_json::Value>,
    error: Option<String>,
}

pub struct Scraper {
    chrome_bin_path: String,
    debug: bool,
    manual: bool,
    mobile: bool,
}

impl Scraper {
    pub fn new() -> Result<Self> {
        NodeJS::check()?;

        let chromium = Chromium::check()?;

        Ok(Scraper {
            chrome_bin_path: chromium.get_bin_path().to_string(),
            debug: false,
            manual: false,
            mobile: false,
        })
    }

    pub fn debug(&mut self) {
        self.debug = true;
    }

    pub fn manual(&mut self) {
        self.manual = true;
    }

    pub fn emulate_mobile(&mut self) {
        self.mobile = true;
    }

    pub fn scrape(&self, url: &str) -> Result<Vec<ScrapeResult>> {
        log::info!("Scraping data from '{}'", url);

        let script_temp_file = get_script_temp_file()?;

        let mut envs = HashMap::new();
        envs.insert("CHROME_BIN_PATH", self.chrome_bin_path.as_str());

        let mut args = vec!["--title", "scraper", &script_temp_file.path, url];

        if self.debug {
            args.push("--debug");
        }

        if self.manual {
            args.push("--manual")
        }

        if self.mobile {
            args.push("--mobile")
        }

        let result = run_command_with_envs("node", args, envs).context("scrape failed")?;

        // TODO handle plain file downloads

        serde_json::from_str(&result).context("failed to parse JSON")
    }
}

#[cfg(test)]
mod tests {
    use anyhow::Result;

    use crate::{ScrapeResult, Scraper};

    fn scrape(url: &str) -> Result<Vec<ScrapeResult>> {
        let scraper = Scraper::new()?;

        scraper.scrape(url)
    }

    #[test]
    #[ignore]
    fn test_scrape_facebook_post() -> Result<()> {
        let result = scrape("https://www.facebook.com/theprodigyofficial/posts/pfbid02XeNwZbYFN8TeXtYrgSCRLPciWpfNWEu3HaUartDe7X5HUH8XGWeXYbHz8wKdREdml")?;

        insta::assert_json_snapshot!(result, {
            "[].data.permalink" => "[permalink]"
        });

        Ok(())
    }

    #[test]
    #[ignore]
    fn test_scrape_facebook_post_with_multiple_images() -> Result<()> {
        let result = scrape("https://www.facebook.com/theprodigyofficial/posts/pfbid0WVYZ4VTe9sddydcCNQGe7NUjajU92iVjM4TQYJgDpo14hy7zfHpQpdH5s2bWsoqul")?;

        insta::assert_json_snapshot!(result, {
            "[].data.permalink" => "[permalink]"
        });

        Ok(())
    }
}
