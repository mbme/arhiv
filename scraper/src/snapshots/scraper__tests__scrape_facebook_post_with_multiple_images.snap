---
source: scraper/src/lib.rs
assertion_line: 123
expression: result
---
[
  {
    "url": "https://www.facebook.com/theprodigyofficial/posts/pfbid0WVYZ4VTe9sddydcCNQGe7NUjajU92iVjM4TQYJgDpo14hy7zfHpQpdH5s2bWsoqul",
    "originalUrl": "https://www.facebook.com/theprodigyofficial/posts/pfbid0WVYZ4VTe9sddydcCNQGe7NUjajU92iVjM4TQYJgDpo14hy7zfHpQpdH5s2bWsoqul",
    "scraperName": "scrapeFBPost",
    "data": {
      "content": "Thanku Brixton for a truly epic night ,,\nThanku to all the people for making this whole tour so massive,\nIt‚Äôs been uplifting, emotional, sonically violent , crowd chaotic ,10,000 degrees‚Ä¶..\nall that shit we luv ,!!\nwe are energised by ur support and luv and we will never forget that . Respect always to the worldwide Prodigy warriors who represent, we luv u all..\nSee u mutha fukkas real soon ‚úäüè¥‚Äç‚ò†Ô∏èüè¥‚Äç‚ò†Ô∏èüè¥‚Äç‚ò†Ô∏èüè¥‚Äç‚ò†Ô∏è\n\n#theprodigy\n#weliveforever\n\nPics by @rhlsngh",
      "date": "24 July at 06:19",
      "dateISO": "2022-07-24T03:19:00.000Z",
      "images": [
        "https://scontent.flwo6-1.fna.fbcdn.net/v/t39.30808-6/294931331_605487010937447_5851313217691047658_n.jpg?stp=dst-jpg_s261x260&_nc_cat=110&ccb=1-7&_nc_sid=8024bb&_nc_ohc=49NMSvD9TvMAX_r_eM6&_nc_ht=scontent.flwo6-1.fna&oh=00_AT9NfO74rdmdiNQ3Uy3TiFxSSVrPtzbYSxCn6CxjLNekqQ&oe=62E42E54",
        "https://scontent.flwo6-1.fna.fbcdn.net/v/t39.30808-6/295516141_605487014270780_7057612721924215012_n.jpg?stp=dst-jpg_s261x260&_nc_cat=105&ccb=1-7&_nc_sid=8024bb&_nc_ohc=NP3YIpRxUFkAX8XNtr4&_nc_oc=AQkXDbRRM9nptb1yZcs6RlOuhim1E2QBTNh99R7VWtjBddyEyVGhoky1OvJVDJyaq5M&_nc_ht=scontent.flwo6-1.fna&oh=00_AT9bgidjwqSIGxRkAW0Y2NDdkUvLIZ_nrja_u58Nv1OCYw&oe=62E52A9E",
        "https://scontent.flwo6-1.fna.fbcdn.net/v/t39.30808-6/295255480_605487007604114_3995477224120169571_n.jpg?stp=dst-jpg_p168x128&_nc_cat=106&ccb=1-7&_nc_sid=8024bb&_nc_ohc=f4G3OZMuebAAX8PQEDf&_nc_ht=scontent.flwo6-1.fna&oh=00_AT9AcRy7EcDwezApksFj2kLlh-IYmhUBB5YbD_Ptyay0AA&oe=62E5708F",
        "https://scontent.flwo6-1.fna.fbcdn.net/v/t39.30808-6/295558398_605487004270781_700684560552675133_n.jpg?stp=dst-jpg_p168x128&_nc_cat=110&ccb=1-7&_nc_sid=8024bb&_nc_ohc=TU_3_uxSQu4AX9OTbp6&_nc_ht=scontent.flwo6-1.fna&oh=00_AT_vJWxxJC7jBGOlGoFnLAMnhxf0zPDDKhRKPE9EzH0j0w&oe=62E4EB83",
        "https://scontent.flwo6-1.fna.fbcdn.net/v/t39.30808-6/294399027_605487017604113_2503223099144554172_n.jpg?stp=dst-jpg_p168x128&_nc_cat=107&ccb=1-7&_nc_sid=8024bb&_nc_ohc=hg5N8hcOJqkAX_Y4SY1&_nc_ht=scontent.flwo6-1.fna&oh=00_AT-Ym0sosXED6HbujyIPGa4xUJdMamyqDtHj-SI5rBy9kg&oe=62E60287"
      ],
      "permalink": "[permalink]"
    },
    "error": null
  }
]
